lib/hllcounter.cc

```diff
@@ -0,0 +1,425 @@
+                break;
+              case 6:
+                v.assign(RAW_BIAS_DATA_6, arr_len(RAW_BIAS_DATA_6));
+                break;
+              case 7:
+                v.assign(RAW_BIAS_DATA_7, arr_len(RAW_BIAS_DATA_7));
+                break;
+              case 8:
+                v.assign(RAW_BIAS_DATA_8, arr_len(RAW_BIAS_DATA_8));
+                break;
+              case 9:
+                v.assign(RAW_BIAS_DATA_9, arr_len(RAW_BIAS_DATA_9));
+                break;
+              case 10:
+                v.assign(RAW_BIAS_DATA_10, arr_len(RAW_BIAS_DATA_10));
+                break;
+              case 11:
+                v.assign(RAW_BIAS_DATA_11, arr_len(RAW_BIAS_DATA_11));
+                break;
+              case 12:
+                v.assign(RAW_BIAS_DATA_12, arr_len(RAW_BIAS_DATA_12));
+                break;
+              case 13:
+                v.assign(RAW_BIAS_DATA_13, arr_len(RAW_BIAS_DATA_13));
+                break;
+              case 14:
+                v.assign(RAW_BIAS_DATA_14, arr_len(RAW_BIAS_DATA_14));
+                break;
+              case 15:
+                v.assign(RAW_BIAS_DATA_15, arr_len(RAW_BIAS_DATA_15));
+                break;
+              case 16:
+                v.assign(RAW_BIAS_DATA_16, arr_len(RAW_BIAS_DATA_16));
+                break;
+              case 17:
+                v.assign(RAW_BIAS_DATA_17, arr_len(RAW_BIAS_DATA_17));
+                break;
+              case 18:
+                v.assign(RAW_BIAS_DATA_18, arr_len(RAW_BIAS_DATA_18));
+                break;
+            }
+            biasData[i] = v;
+        }
+    }
+}
+
+double get_threshold(int p)
+{
+    return THRESHOLD_DATA[p - 4];
+}
+
+vector<int> get_nearest_neighbors(double E, vector<double> estimate)
+{
+    vector< pair<double,int> > distance_map;
+    vector<int> nearest;
+
+    int i = 0;
+    for (vector<double>::iterator it = estimate.begin();
+         it != estimate.end();
+         ++it) {
+      std::pair<double, int> p(pow(E - *it, 2.0), i);
+      distance_map.push_back(p);
+      i++;
+    }
+
+    sort(distance_map.begin(), distance_map.end());
+
+    for(int k=0; k < 6; k++) {
+        nearest.push_back(distance_map[k].second);
+    }
+
+    return nearest;
+}
+
+double estimate_bias(double E, int p)
+{
+    vector<double> bias = biasData[p];
+    vector<double> raw_estimate = rawEstimateData[p];
+
+    vector<int> nearest = get_nearest_neighbors(E, raw_estimate);
+    double estimate = 0.0;
+
+    for (vector<int>::iterator it = nearest.begin();
+         it != nearest.end();
+         ++it) {
+      estimate += bias[*it];
+    }
+    return estimate / nearest.size();
+}
+
+double ep_sum(double acc, int b)
+{
+    return acc += pow(2.0, float(-b));
+}
+
+int get_rho(HashIntoType w, int max_width)
+{
+    int rho = max_width - floor(log2(w)); /* TODO find last bit set */
+
+    if (rho <= 0) {
+        return -1;    // TODO: treat error, w overflow
```

Comment:
I will add issues for remaining TODO in code and for the list in the end of this discussion, tag them as HLL improvements and assign to me.